src/plenoptic/simulate/models/portilla_simoncelli.py

```diff
@@ -82,934 +92,849 @@ def __init__(
-        )
-        rep["auto_correlation_magnitude"] = vec[n_filled : (n_filled + nn)].unflatten(
-            0,
-            (
-                self.spatial_corr_width,
-                self.spatial_corr_width,
-                self.n_scales,
-                self.n_orientations,
-            ),
-        )
-        n_filled += nn
-
-        # skew_reconstructed & kurtosis_reconstructed
-        nn = self.n_scales + 1
-        rep["skew_reconstructed"] = vec[n_filled : (n_filled + nn)]
-        n_filled += nn
-
-        rep["kurtosis_reconstructed"] = vec[n_filled : (n_filled + nn)]
-        n_filled += nn
-
-        # auto_correlation_reconstructed
-        nn = self.spatial_corr_width * self.spatial_corr_width * (self.n_scales + 1)
-        rep["auto_correlation_reconstructed"] = vec[
-            n_filled : (n_filled + nn)
-        ].unflatten(
-            0, (self.spatial_corr_width, self.spatial_corr_width, self.n_scales + 1)
-        )
-        n_filled += nn
-
-        if self.use_true_correlations:
-            nn = self.n_scales + 1
-            rep["std_reconstructed"] = vec[n_filled : (n_filled + nn)]
-            n_filled += nn
-
-        # cross_orientation_correlation_magnitude
-        nn = self.n_orientations * self.n_orientations * (self.n_scales + 1)
-        rep["cross_orientation_correlation_magnitude"] = vec[
-            n_filled : (n_filled + nn)
-        ].unflatten(0, (self.n_orientations, self.n_orientations, self.n_scales + 1))
-        n_filled += nn
-
-        # cross_scale_correlation_magnitude
-        nn = self.n_orientations * self.n_orientations * self.n_scales
-        rep["cross_scale_correlation_magnitude"] = vec[
-            n_filled : (n_filled + nn)
-        ].unflatten(0, (self.n_orientations, self.n_orientations, self.n_scales))
-        n_filled += nn
-
-        # cross_orientation_correlation_real
-        nn = (
-            max(2 * self.n_orientations, 5)
-            * max(2 * self.n_orientations, 5)
-            * (self.n_scales + 1)
-        )
-        rep["cross_orientation_correlation_real"] = vec[
-            n_filled : (n_filled + nn)
-        ].unflatten(
-            0,
-            (
-                max(2 * self.n_orientations, 5),
-                max(2 * self.n_orientations, 5),
-                self.n_scales + 1,
-            ),
-        )
-        n_filled += nn
+        rep = einops.pack(list(representation_dict.values()), 'b c *')[0]
+        # then get rid of all the nans / unnecessary stats
+        return rep.index_select(-1, self._necessary_stats_mask)
 
-        # cross_scale_correlation_real
-        nn = 2 * self.n_orientations * max(2 * self.n_orientations, 5) * self.n_scales
-        rep["cross_scale_correlation_real"] = vec[n_filled : (n_filled + nn)].unflatten(
-            0, (2 * self.n_orientations, max(2 * self.n_orientations, 5), self.n_scales)
-        )
-        n_filled += nn
+    def convert_to_dict(self, representation_vector: Tensor) -> OrderedDict:
```

Comment:
I like this in general, though for this to work, `dict_to_tensor` has to be called before `tensor_to_dict`, which I don't think is great. metadata could be passed at initialization to avoid this.

also, for PortillaSimoncelli this `tensor_to_dict` won't work because of the "missing statistics": we throw away redudant statistics so that, for example, there are about half the values corresponding to the autocorrelation in the tensor (that's what the NaNs are doing in the current `convert_to_dict`). 

In more detail: if we have a 5x5 autocorrelation matrix, our tensor only contains 12 of these values (the lower triangle plus the first two values of the diagonal), as the central value is always 1 and the others are redundant. In order to convert this back to the dictionary, we create a 5x5 array of NaNs and need to know where to insert those 12 values, which comes from the True values in `_necessary_stats_dict` (so carrying more than just the shape information).

Does that make sense?

Reply:
so the more general solution would still need access to `_necessary_stats_dict` in  `tensor_to_dict` . and then I'm not sure it's worth it.

Reply:
I see, that makes sense to me; If there is no way to make it general, I agree it's not worth the effort